File: 剑指offer/32.3.从上到下打印二叉树.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrder(self, root: TreeNode) -> List[int]:
        ans = []

        if not root:
            return []

        queue = [root]
        order = 0
        while queue:
            tmp = []
            for _ in range(len(queue)):
                cur = queue.pop(0)
                tmp.append(cur.val)

                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)

            if order % 2 == 0:
                ans.append(tmp)
            else:
                ans.append(tmp[::-1])
            order += 1

        return ans

File: 剑指offer/32.3.从上到下打印二叉树.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root)
    {
        vector<vector<int>> ans;
        int direction = 0;

        if (!root) {
            return {};
        }

        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int sz = q.size();
            vector<int> tmp;
            for (int i = 0; i < sz; i++) {
                TreeNode* cur = q.front();
                q.pop();
                tmp.push_back(cur->val);

                if (cur->left) {
                    q.push(cur->left);
                }
                if (cur->right) {
                    q.push(cur->right);
                }
            }
            if (direction == 1) {
                reverse(tmp.begin(), tmp.end());
            }
            direction = (direction + 1) % 2;
            ans.push_back(tmp);
        }

        return ans;
    }
};